File: 967-numbers-with-same-consecutive-differences/967-numbers-with-same-consecutive-differences.cpp
class Solution {
public:
    
    int n,k;
    
    void DFS(int i, int d, int &pth, vector<int> &ret)
    {
        pth=pth*10+d;
        
        if(i==n-1)
        {
         ret.push_back(pth);
         pth=pth/10;
         return;
        }
        if(d-k>=0 && d-k <=9)
        {
            DFS(i+1,d-k,pth, ret);
        }
        if(d+k>=0 && d+k<=9 && k!=0)
        {
            DFS(i+1,d+k,pth,ret);
        }
        pth=pth/10;
    }
    
    vector<int> numsSameConsecDiff(int n1, int k1) {
        
        n=n1; k=k1; int pth=0;
        vector<int> ret;
        
        for(int i=1;i<=9;i++) DFS(0,i,pth,ret);
        
        return(ret);
        
    }
};

File: 967-numbers-with-same-consecutive-differences/967-numbers-with-same-consecutive-differences.java
class Solution {
    public int[] numsSameConsecDiff(int n, int k) {
        List<Integer> numbers = new ArrayList<>();
        for(int i = 1;  i <= 9; i++) {
            dfs(i, n, k, numbers, 0, 0);
        }
        
        int[] result = new int[numbers.size()];
        int j = 0;
        
        for(Integer i : numbers) {
            result[j++] = i;
        }
        
        return result;
    }
    
    void dfs(int num, int n, int k, List<Integer> numbers, int currentNumber, int currentLength) {

        currentNumber += num;
        currentLength += 1;
        
        if(currentLength == n) {
            numbers.add(currentNumber);
            return;
        }
        
        if(num + k < 10) {
            dfs(num + k, n, k, numbers, currentNumber * 10, currentLength);
        }
        
        if(k != 0 && num - k >= 0) {
            dfs(num - k, n, k, numbers, currentNumber * 10, currentLength);
        }
    }
}